pkg/kclient/all.go

```diff
@@ -56,17 +56,23 @@ func getAllResources(client dynamic.Interface, apis []apiResource, ns string, se
 			return nil
 		})
 	}
+	klog.V(2).Infof("fired up all goroutines to query APIs")
+
+	errChan := make(chan error)
+	go func() {
+		errChan <- group.Wait()
+		klog.V(2).Infof("all goroutines have returned in %v", time.Since(start))
+		close(outChan) // because we are ranging over the channel below
+	}()
 
 	go func() {
 		for val := range outChan {
 			out = append(out, val...)
 		}
 	}()
 
-	klog.V(2).Infof("fired up all goroutines to query APIs")
-	klog.V(2).Infof("all goroutines have returned in %v", time.Since(start))
 	klog.V(2).Infof("query result: objects=%d", len(out))
-	return out, group.Wait()
+	return out, <-errChan
```

Comment:
Okay, with the `for-range` in a separate goroutine, we avoid making `errChan` a buffered channel. This function will return as soon as a value is read from `errChan`, but I think we still cannot guarantee that the `for-range` has finished iterating at this point. So we cannot be sure that we have read everything from `outChan`, no?

Reply:
Why would we want to read everything from `outChan` if there was an error?

Reply:
I meant that even if there was no error, `out` would contain everything only after `outChan` is closed, which, in your changes, happens after the result of `group.Wait()` has been written to `errChan`. But at the same time, `<-errChan` would no longer be blocking, but we are not sure `outChan` has been closed yet when returning.
In other words, `return out, <-errChan` could return before we are done iterating over `outChan`, I think.

Reply:
If you are iterating over `outChan` in a separate goroutine, then maybe the calls to `group.Wait()` and `close(outChan)` could be done right in the function itself, not another goroutine.. I think this would be useful if we were `for-range`'ing in the main function itself.

```diff
-       errChan := make(chan error)
-       go func() {
-               errChan <- group.Wait()
-               klog.V(2).Infof("all goroutines have returned in %v", time.Since(start))
-               close(outChan) // because we are ranging over the channel below
-       }()
-
        go func() {
                for val := range outChan {
                        out = append(out, val...)
                }
        }()
 
+       err := group.Wait()
+       klog.V(2).Infof("all goroutines have returned in %v", time.Since(start))
+       close(outChan)
+
        klog.V(2).Infof("query result: objects=%d", len(out))
-       return out, <-errChan
+       return out, err

```

Reply:
> I meant that even if there was no error, `out` would contain everything only after `outChan` is closed, which, in your changes, happens after the result of `group.Wait()` has been written to `errChan`. But at the same time, `<-errChan` would no longer be blocking, but we are not sure `outChan` has been closed yet when returning.
> In other words, `return out, <-errChan` could return before we are done iterating over `outChan`, I think.

I still don't find the answer to - Why would we want to read everything from `outChan` if there was an error?

> I meant that even if there was no error, `out` would contain everything only after `outChan` is closed, which, in your changes, happens after the result of `group.Wait()` has been written to `errChan`.

`out` already has it. It's just that the goroutine ranging over `outChan` keeps ranging over the channel. That's how, in spite of https://github.com/redhat-developer/odo/pull/5973#discussion_r933315813, we get the expected result with `odo logs` and `odo logs --follow`. Besides that, there's no other time when we could close the `outChan` - we need to wait til `group.Wait()` is done.

> But at the same time, `<-errChan` would no longer be blocking, but we are not sure `outChan` has been closed yet when returning.

And why do we want to be sure to close the `outChan`? We encountered an error, and we want that to pass on to the user as soon as doable. Are you concerned about the leaky aspect of it? Since we are stopping things as soon as we encounter an error, it's not a problem for us. If we were running an API server, we would have to be bothered about closing the channel.

Reply:
> > I meant that even if there was no error, `out` would contain everything only after `outChan` is closed, which, in your changes, happens after the result of `group.Wait()` has been written to `errChan`. But at the same time, `<-errChan` would no longer be blocking, but we are not sure `outChan` has been closed yet when returning.
> > In other words, `return out, <-errChan` could return before we are done iterating over `outChan`, I think.
> 
> I still don't find the answer to - Why would we want to read everything from `outChan` if there was an error?
> 
> > I meant that even if there was no error, `out` would contain everything only after `outChan` is closed, which, in your changes, happens after the result of `group.Wait()` has been written to `errChan`.
> 
> `out` already has it. It's just that the goroutine ranging over `outChan` keeps ranging over the channel. That's how, in spite of [#5973 (comment)](https://github.com/redhat-developer/odo/pull/5973#discussion_r933315813), we get the expected result with `odo logs` and `odo logs --follow`. Besides that, there's no other time when we could close the `outChan` - we need to wait til `group.Wait()` is done.
> 
> > But at the same time, `<-errChan` would no longer be blocking, but we are not sure `outChan` has been closed yet when returning.
> 
> And why do we want to be sure to close the `outChan`? We encountered an error, and we want that to pass on to the user as soon as doable. Are you concerned about the leaky aspect of it? Since we are stopping things as soon as we encounter an error, it's not a problem for us. If we were running an API server, we would have to be bothered about closing the channel.

I agree with you about the case where we encounter an error. I was mostly thinking about the other case, where `group.Wait()` would not return any error (`nil` error written to `errChan` and read from it): my point was that with the previous changes, `out` could have been filled partially if `outChan` was not closed before the function returns.
But overall yes, my concern is about not leaking goroutines, but with long-running processes in mind, like `odo logs --follow` (or any other long-running command that could use this function).

Anyways, thanks for the discussion. Your new changes now look good to me.